python-checks/src/main/java/org/sonar/python/checks/cdk/IamPolicyPublicAccessCheck.java

```diff
@@ -0,0 +1,172 @@
+          sensitivePrincipal.addIssue(ISSUE_MESSAGE);
+        }
+      });
+  }
+
+  private static void checkPolicyStatementJson(SubscriptionContext subscriptionContext, CallExpression callExpression) {
+    if (callExpression.arguments().isEmpty()) {
+      return;
+    }
+
+    getArgument(subscriptionContext, callExpression, "obj", 0)
+      .flatMap(CdkUtils::getDictionary)
+      .ifPresent(dictionaryLiteral -> checkJsonEntry(subscriptionContext, dictionaryLiteral));
+  }
+
+  private static void checkPolicyDocumentJson(SubscriptionContext subscriptionContext, CallExpression callExpression) {
+    Optional<DictionaryLiteral> dictionaryLiteral = getArgument(subscriptionContext, callExpression, "obj", 0)
+      .flatMap(CdkUtils::getDictionary);
+
+    if (dictionaryLiteral.isEmpty()) {
+      return;
+    }
+
+    Optional<ListLiteral> statement = CdkUtils.getDictionaryPair(subscriptionContext, dictionaryLiteral.get(), "Statement")
+      .flatMap(resolvedKeyValuePair -> getList(resolvedKeyValuePair.value));
+
+    if (statement.isEmpty()) {
+      return;
+    }
+
+    getListElements(subscriptionContext, statement.get())
+      .stream()
+      .map(CdkUtils::getDictionary)
+      .flatMap(Optional::stream)
+      .forEach(innerDict -> checkJsonEntry(subscriptionContext, innerDict));
+  }
+
+  private static void checkJsonEntry(SubscriptionContext subscriptionContext, DictionaryLiteral dictionaryLiteral) {
+    List<CdkUtils.ResolvedKeyValuePair> resolvedKeyValuePairs = resolveDictionary(subscriptionContext, dictionaryLiteral);
+
+    Optional<CdkUtils.ResolvedKeyValuePair> effect = getDictionaryPair(resolvedKeyValuePairs, "Effect");
+    Expression effectSetting = effect.flatMap(expressionFlow -> expressionFlow.value.getExpression(isString("Allow"))).orElse(null);
+
+    if (effect.isPresent() && effectSetting == null) {
+      // "Effect" is defined, and it is not equal to "Allow"
+      return;
+    }
+
+    // Given an entry under the key "Principal", we want to raise an issue if the value is either
+    //  1) string literal "*",
+    //  2) dictionary in the format of { "AWS": "*" } or { "AWS": ["99999", ..., "*", ....] }
+    getDictionaryPair(resolvedKeyValuePairs, "Principal").map(pair -> pair.value).ifPresent(principal -> {
+      raiseIssueIf(principal, isString("*"), effectSetting);
+
+      getDictionary(principal)
+        .flatMap(innerDict -> getDictionaryPair(subscriptionContext, innerDict, "AWS"))
+        .ifPresent(aws -> {
+          // Check if the principal value is set to { "AWS": "*" }
+          raiseIssueIf(aws.value, isString("*"), effectSetting);
+
+          // Check if the principal value is set to { "AWS": ["99999", ..., "*", ....] }
+          getListExpression(aws.value)
+            .ifPresent(listLiteral ->
+              getListElements(subscriptionContext, listLiteral)
+                .forEach(principalElement -> raiseIssueIf(principalElement, isString("*"), effectSetting)));
+        });
+    });
```

Comment:
This a good example of the overuse of Optionals. TBH we can keep it for now and refactor it in the next sprint. WDYT?

Reply:
Makes sense to me, let's talk it in the clean-up effort :+1: